### PR TITLE
Convert WebGPU compiler target into a plugin.

### PR DESCRIPTION
**Pull request by ScottTodd**
Progress on https://github.com/openxla/iree/issues/15468

I didn't do anything fancy with includes or namespaces for this, just relative includes with `./` (same as target/CUDA).
* Unit test is at `iree/compiler/plugins/target/WebGPU/test/smoketest.mlir.test`
* Unit tests for input plugins are at e.g. `tosa-iree/InputConversion/test/auto_input_conversion.mlir.test`

I also moved the deps that are unique to the WebGPU target into the plugin, keeping the root `CMakeLists.txt` and shared `build_tools/third_party/` directories cleaner.

### CODE REVIEW

**Review by antiagainst — APPROVED** [review]

File: compiler/plugins/target/WebGPU/CMakeLists.txt [hunk 1]
```diff
@@ -4,18 +4,20 @@
...
+iree_compiler_register_plugin(
+  PLUGIN_ID
+    hal_target_webgpu
+  TARGET
+    ::WebGPU
+)
+
 iree_cc_library(
   NAME
     WebGPU
```
Comment (ScottTodd) [inline compiler/plugins/target/WebGPU/CMakeLists.txt:36]: We can also move the supporting build system configuration into this plugin directory, similar to how Torch does this: https://github.com/openxla/iree/blob/main/compiler/plugins/input/Torch/torch-mlir/CMakeLists.txt

These are the pieces that can move:

* https://github.com/openxla/iree/blob/main/build_tools/third_party/tint/CMakeLists.txt
* https://github.com/openxla/iree/blob/ac9548469e8e46f8ad6ea29c1c7144a774ebba19/CMakeLists.txt#L930-L946

**Comment by ScottTodd** [comment]
PTAL. I reworked some dependencies.

File: compiler/plugins/target/WebGPU/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# or they can be updated independently
+set(IREE_TINT_TAG          "fdb8787e9c1b79770bd98a8faf37fbe48a3077a4")  # 2023-03-06
+set(IREE_SPIRV_HEADERS_TAG "b730938c033ede3572b660ab019b438509ba24d9")  # 2023-08-10
+set(IREE_SPIRV_TOOLS_TAG   "43b8886490eb6af81fc61e0ff071c51a922af864")  # 2023-08-11
+message(STATUS "Configuring WebGPU target deps (SPIRV-Headers, SPIRV-Tools, Tint)")
+list(APPEND CMAKE_MESSAGE_INDENT "  ")
+add_subdirectory(spirv-headers EXCLUDE_FROM_ALL)
+add_subdirectory(spirv-tools EXCLUDE_FROM_ALL)
+add_subdirectory(tint EXCLUDE_FROM_ALL)
+list(POP_BACK CMAKE_MESSAGE_INDENT)
```
Comment (ScottTodd) [inline compiler/plugins/target/WebGPU/CMakeLists.txt:23]: Sample logs: https://github.com/openxla/iree/actions/runs/6895343970/job/18759132318?pr=15612#step:4:514

```
-- Registering static linked compiler plugin 'hal_target_cuda' (iree::compiler::plugins::target::CUDA::CUDA)
-- Configuring WebGPU target deps (SPIRV-Headers, SPIRV-Tools, Tint)
--   Tint build samples: OFF
--   Tint build docs: OFF
--   Tint build docs with warn as error: OFF
--   Tint build SPIR-V reader: ON
--   Tint build WGSL reader: OFF
--   Tint build GLSL writer: OFF
--   Tint build HLSL writer: OFF
--   Tint build MSL writer: OFF
--   Tint build SPIR-V writer: OFF
--   Tint build WGSL writer: ON
--   Tint build IR: ON
--   Tint build fuzzers: OFF
--   Tint build SPIRV-Tools fuzzer: OFF
--   Tint build AST fuzzer: OFF
--   Tint build regex fuzzer: OFF
--   Tint build benchmarks: OFF
--   Tint build tests: OFF
--   Tint build with ASAN: OFF
--   Tint build with MSAN: OFF
--   Tint build with UBSAN: OFF
--   Tint build checking [chromium-style]: OFF
--   Tint build remote-compile tool: OFF
--   Using python3
--   Found PythonInterp: /usr/bin/python3 (found suitable version "3.9.5", minimum required is "3") 
-- Registering static linked compiler plugin 'hal_target_webgpu' (iree::compiler::plugins::target::WebGPU::WebGPU)
```

(Similar indentation changes are in https://github.com/openxla/iree/pull/15596)

Comment (antiagainst) [reply compiler/plugins/target/WebGPU/CMakeLists.txt:23]: Neat! I've learnt a trick. :)

**Review by antiagainst — APPROVED** [review]

**Pull request merged**